### PR TITLE
Docstring improvements (type annotations, `data_format` and `weights` consistency)

### PR DESCRIPTION
**Pull request by SamuelMarks**
Shouldn't be anything controversial here. PR might be too large, so if you want it split to one PR per file just say the word.

The overall goal here is to expose Keras as a fully-typed SDK, CLI, GUI, series of SQL tables. This will enable a number of use-cases, most interesting (IMHO) is being able to automatically compare / mix-and-match different parts of the ontology.

E.g., mix-and-match optimisers, loss functions, applications (transfer-learning models). Automatically. For example, consider AutoML, hyperparameter and parameter optimisation use-cases.

PS: Future work, in terms of PRs to Keras, include further improvements to
- docstrings 
- type constraints (e.g., numerical ranges of `integer > 0`, or `float in range [0, 1]` [or a concise [interval](https://en.wikipedia.org/wiki/Interval_(mathematics)) syntax]).
- automatic switching from types in docstrings to types in source (optionally removing `Defaults to` from docstring); e.g., my Python compiler can already do this.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR -- unfortunately this is not a good fit. Docstrings are intended to be markdown-rendereable and human-readable. They are not an ontology. We should avoid symbolic type annotations, and further we should avoid lists in argument descriptions because we often can't render them correctly.

> The overall goal here is to expose Keras as a fully-typed SDK, CLI, GUI, series of SQL tables

This could be done via a separate data structure. However the challenge will be to keep it up to date with the code (this problem also exists if you use docstrings to store the info, of course). Perhaps it would be better to automatically generate such an ontology from the source code itself.

**Comment by SamuelMarks** [comment]
@fchollet Yes that is exactly the idea; automatically generating the ontology from the source-code itself.

Unfortunately your current SDK isn't amenable to this. So I am sending you PRs to make it trivial for such a parser to deterministically determine the acceptable inputs (and later; their ranges).

(and later, with a single command, convert all docstring types to source-code literal type hints, or in a first/third party generated `.pyi` file)

The only reason I chose the `-` syntax for lists to specify sub-elements in your docstrings is that I saw them—albeit inconsistently—existent throughout your codebase.

Happy to choose some other syntax. There's a rather ugly syntax proposed for treating docstrings as actual code ([PEP727](https://peps.python.org/pep-0727/)).

(there's also the general critique of having differently detailed descriptions for the same argument name; like `data_format`)

Regardless making docstrings consistent throughout your codebase shouldn't be controversial. Just tell me which syntax you want to follow, and I'll consistently apply that style throughout, and expand my docstring parser/emitter with support.

Thanks for reconsidering
